app/models/story_parser.rb

```diff
@@ -421,8 +421,12 @@ def parse_author_common(email, name)
       name = name.to_ascii.gsub(/[^\w[ \-@.]]/u, "")
       external_author = ExternalAuthor.find_or_create_by(email: email)
       external_author_name = external_author.default_name
+
+      # if the name and email don't exist in the DB tables, add it
       unless name.blank?
-        external_author_name = ExternalAuthorName.where(name: name, external_author_id: external_author.id).first ||
+        raise Error, external_author.errors.full_messages.join(" ") if external_author.invalid?
```

Comment:
I think this line is in the wrong place. If someone enters `!!!!!` as the name, and also uses an invalid email address, then the exclamation points in the name will be stripped down so that `name` is blank. So even though it's an invalid email, I don't think it will reach this line and show the error. I'm pretty sure this line needs to be outside of the `unless name.blank?` block.

Reply:
Thank you for your comment. In that scenario (for the case where the name is blank and the email is invalid), should the error just say "invalid email"? Or should it alert about the name *and* the email?

Reply:
It'd be more difficult to handle that case, since the `ExternalAuthor` currently doesn't get created if either the `name` or the `email` is blank, so it never runs through the validator. Since it's not producing a 500 error, I'd be tempted to let it slide for the moment, but someone who's more familiar with the Open Doors committee might have a better answer for you.

(Really, I think the entire function should be rewritten, so that there are proper error messages for everything instead of silently fixing invalid names, and it's easier to follow the code. But IDK if that's in scope, and IDK if that's what Open Doors wants.)

Reply:
I also reorganized ``parser_author_common`` in a way that makes sense to me. And to avoid breaking something, I added tests and confirmed the behavior didn't change (as far as I can see) except for the email address validation, too.

Reply:
If you're going to rewrite it for clarity reasons, I think it'd be cleaner to consolidate most of the name-related code. As is, `redacted_name` is set, and then there are a bunch of lines involving errors & emails, and then `redacted_name` is finally used. But there's no reason for it to be set that early, and it's easier to keep track of the variable's purpose if it's set right before it's used.

Similarly, the two different possible values of `external_author_name` are set in two different `if` blocks, so it's hard to see at a glance what the function is supposed to return. It'd be cleaner to set it equal to `external_author.default_name` in the `else` of the `if redacted_name.present?` block.

With some other cleanup, that could get you something like this:
```ruby
  def parse_author_common(email, name)
    errors = []

    errors << "No author name specified" if name.blank?

    if email.present?
      external_author = ExternalAuthor.find_or_create_by(email: email)
      errors += external_author.errors.full_messages
    else
      errors << "No author email specified"
    end

    raise Error, errors.join("\n") if errors.present?

    # convert to ASCII and strip out invalid characters (everything except alphanumeric characters, _, @ and -)
    redacted_name = name.to_ascii.gsub(/[^\w[ \-@.]]/u, "")
    if redacted_name.present?
      external_author.names.find_or_create_by(name: redacted_name)
    else
      external_author.default_name
    end
  end
```
Ideally, it'd also be nice to move some or all of the error-checking code here into the models, but that's almost certainly out of scope.

Reply:
Thanks a lot for the rewrite. Do you think  ``.save`` is unneccesary? I thought the method makes persistent changes to the databases so that author names can be retrieved by querying emails later. (Or maybe it's implicitly saved somehow?)

Reply:
`find_or_create_by` calls `create` [under the hood](https://github.com/rails/rails/blob/593893c901f87b4ed205751f72df41519b4d2da3/activerecord/lib/active_record/encryption/extended_deterministic_queries.rb#L94-L96), and `create` calls `save` [under the hood](https://github.com/rails/rails/blob/593893c901f87b4ed205751f72df41519b4d2da3/activerecord/lib/active_record/persistence.rb#L33-L41). So both the external author and the external author name are getting saved to the database if they don't already exist, because they're both fetched/created with `find_or_create_by` in the modified code.

Reply:
Understood. Thanks for explaining.